src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -448,24 +448,34 @@
     (shutdown-on-error [this svc-id f] (shutdown-on-error* shutdown-reason-promise app-context svc-id f))
     (shutdown-on-error [this svc-id f on-error] (shutdown-on-error* shutdown-reason-promise app-context svc-id f on-error))))
 
-(schema/defn ^:always-validate shutdown!
+(schema/defn ^:always-validate shutdown! :- [Throwable]
   "Perform shutdown calling the `stop` lifecycle function on each service,
-   in reverse order (to account for dependency relationships)."
+   in reverse order (to account for dependency relationships).
+   Returns collection of exceptions thrown during shutdown sequence execution."
   [app-context :- (schema/atom a/TrapperkeeperAppContext)]
   (log/info (i18n/trs "Beginning shutdown sequence"))
   (let [{:keys [ordered-services shutdown-channel lifecycle-worker]} @app-context
-        shutdown-fn (fn [] (doseq [[service-id s] (reverse ordered-services)]
-                             (try
-                               (run-lifecycle-fn! app-context s/stop "stop" service-id s)
-                               (catch Exception e
-                                 (log/error e (i18n/trs "Encountered error during shutdown sequence"))))))]
+        result-chan (async/promise-chan (map #(filter (partial not= :ok) %)))
+        shutdown-fn (fn [] (let [results
+                                 (doall
+                                  (map (fn [[service-id s]]
+                                         (try
+                                           (run-lifecycle-fn! app-context s/stop "stop" service-id s)
+                                           :ok
```

Comment:
Style: Perhaps we could return `nil` here and use `keep` instead of `map`. Maybe then we could remove the `promise-chan` transducer.

Reply:
I think you are right, it looks much cleaner. Actually I begin with that idea but later my thought drifted towards "results-chan" instead of "errors-chan" as a holder for stuses not only for "failed to stop" services but also for all others, like "service finished but there is a list of pending jobs"